modules/economy.js

```diff
@@ -1,293 +1,235 @@
-      let propertiesString = ''
+    const memberInputRequirements = properties
+    let memberInput
+    if (member.inputs) memberInput = member.inputs.split(',')
+    const memberInputOrganized = {}
+    const missingInputs = []
 
-      if (userInputRequirements && userInputRequirements.length > 0) {
-        if (item.userInput) {
-          userInputRequirements.forEach(userInputRequirement => {
-            userInputs[userInputRequirement] = 1
-          })
-          item.userInput.forEach(property => {
-            property = property.split(':')
-            if (userInputs[property[0].trim()] && property[1]) {
-              userInputs[property[0]] = property[1].trim()
-            }
-          })
-          Object.keys(userInputs).forEach(userInput => {
-            if (!userInputs[userInput] || userInputs[userInput] === 1) {
-              delete userInputs[userInput]
-              propertiesString += ` ${userInput}`
-            }
-          })
-          if (propertiesString) {
-            callback(i18n(guild.locale, 'BUYPRODUCT_MISSING_PROPERTY', { PROPERTY: propertiesString }), false)
+    if (Array.isArray(memberInputRequirements) && memberInputRequirements.length >= 0 && Array.isArray(memberInput)) {
+      memberInputRequirements.forEach(userInputRequirement => {
+        memberInputOrganized[userInputRequirement] = null
+      })
+      memberInput.forEach(input => {
+        const getPropertyAndItsValue = input.split(':')
+        if (memberInputRequirements[getPropertyAndItsValue[0]]) memberInputRequirements[getPropertyAndItsValue[0]] = getPropertyAndItsValue[1]
+      })
+      Object.keys(memberInputOrganized).forEach(input => {
+        if (memberInputOrganized[input] === null) {
+          delete memberInputRequirements[input]
+          missingInputs.push(input)
+        }
+      })
+      if (Array.isArray(missingInputs) && missingInputs.length > 0) return { code: 'ECO_BU05', missingInputs: missingInputs }
+    }
+
+    if (action && Object.prototype.hasOwnProperty.call(action, 'type')) {
+      switch (action.type) {
+        case 'sendMessage': {
+          if (Object.prototype.hasOwnProperty.call(action, 'message') && Object.prototype.hasOwnProperty.call(action, 'channel')) {
+            guild.channels.fetch(action.channel).then(channel => {
+              if (channel) {
+                channel.send(StringPlaceholder(action.message, memberInputOrganized, { before: '#', after: '#' }) || 'Nothing')
+              } else {
+                throw new Error('ECO_ATI05')
+              }
+            })
           } else {
-            module.exports.addItemToUser(client, member, guild, item.productId, item.productMeta.singlebuy, (status) => {
-              if (status === 'error') {
-                callback(i18n(guild.locale, 'BUYPRODUCT_ALREADYOWN', { PRODUCT: item.productName }), false)
+            throw new Error('ECO_ATI07')
+          }
+          break
+        }
+        case 'giveRole': {
```

Comment:
Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/alexandrumuntenas/pingu/pullRequest?prid=8657799)
```suggestion
                    case "giveRole": {
```